app/src/parser_impl.c

```diff
@@ -703,51 +705,55 @@ uint8_t _countArgumentItems(const flow_argument_list_t *v, uint8_t argumentIndex
     CHECK_PARSER_ERR(json_matchKeyValue(&parsedJson, 0, (char *) "Array", JSMN_ARRAY, &internalTokenElementIdx));
     uint16_t arrayTokenCount;
     CHECK_PARSER_ERR(array_get_element_count(&parsedJson, internalTokenElementIdx, &arrayTokenCount));
-    if (arrayTokenCount > 64) {
+    if (arrayTokenCount > max_number_of_items) {
         return PARSER_UNEXPECTED_NUMBER_ITEMS;
     }
 
-    return arrayTokenCount;
+    *number_of_items = arrayTokenCount;
+    return PARSER_OK;
 }
 
-
-uint8_t _countArgumentOptionalItems(const flow_argument_list_t *v, uint8_t argumentIndex) {
+//if Optional is null, number_of_items is set to 1 as one screen is needed to dispay "None"
+parser_error_t _countArgumentOptionalItems(const flow_argument_list_t *v, uint8_t argumentIndex, 
+                                           uint8_t max_number_of_items, uint8_t *number_of_items) {
+    *number_of_items = 0;
     parsed_json_t parsedJson = {false};
 
     if (argumentIndex >= v->argCount) {
-        return 0;
+        return PARSER_UNEXPECTED_FIELD;
     }
 
     const parser_context_t argCtx = v->argCtx[argumentIndex];
-    parser_error_t err = json_parse(&parsedJson, (char *) argCtx.buffer, argCtx.bufferLen);
-    if (err != PARSER_OK) {
-        return 0;
-    }
+    CHECK_PARSER_ERR(json_parse(&parsedJson, (char *) argCtx.buffer, argCtx.bufferLen));
 
-    // Get numnber of items
     uint16_t internalTokenElementIdx;
-    err = json_matchOptionalArray(&parsedJson, 0, &internalTokenElementIdx);
-    if (err != PARSER_OK) {
-        return 0;
-    }
+    CHECK_PARSER_ERR(json_matchOptionalArray(&parsedJson, 0, &internalTokenElementIdx));
     if (internalTokenElementIdx == JSON_MATCH_VALUE_IDX_NONE) {
-        return 1;
+        *number_of_items = 1;
+        return PARSER_OK;
     }
     
+    // Get numnber of items
     uint16_t arrayTokenCount;
-    err = array_get_element_count(&parsedJson, internalTokenElementIdx, &arrayTokenCount);
-    if (err != PARSER_OK || arrayTokenCount > 64) {
-        return 0;
+    CHECK_PARSER_ERR(array_get_element_count(&parsedJson, internalTokenElementIdx, &arrayTokenCount));
+    if (arrayTokenCount > max_number_of_items) {
+        return PARSER_UNEXPECTED_NUMBER_ITEMS;
     }
-    return arrayTokenCount;
+
+    *number_of_items = arrayTokenCount;
+    return PARSER_OK;
 }
 
+//this function returns incorrect value on parser error, checks for parser errors are performed in next 
```

Comment:
We should document the flow of functions - how do we guarantee that the next call actually happens? Would the device start displaying items? What are the guarantees of the failing functions? Could they modify their arguments to some non-safe values before returning an error?

Reply:
Even better, why don't we just propagate the error?
Parent function is ready to return an error:
```
parser_error_t parser_getNumItems(const parser_context_t *ctx, uint8_t *num_items) {
    *num_items = _getNumItems(ctx, &parser_tx_obj);
    return PARSER_OK;
}
```